java/java.source.base/src/org/netbeans/modules/java/source/indexing/JavaCustomIndexer.java

```diff
@@ -147,6 +151,76 @@ protected void index(final Iterable<? extends Indexable> files, final Context co
+                        break;
+                    }
+                    binaryRootsFo[at++] = ufo;
+                    Enumeration<? extends FileObject> en = ufo.getChildren(true);
+                    while (en.hasMoreElements()) {
+                        FileObject ch = en.nextElement();
+                        long modified = ch.lastModified().getTime();
+                        if (newestFile == null || newestFile < modified) {
+                            newestFile = modified;
+                        }
+                    }
+                }
+
+                boolean binariesAreNewer = true;
+                for (Indexable index : files) {
+                    if (!binariesAreNewer) {
+                        break;
+                    }
+                    FileObject fo = context.getRoot().getFileObject(index.getRelativePath());
+                    if (newestFile == null || fo == null || fo.lastModified().getTime() > newestFile) {
+                        binariesAreNewer = false;
+                        break;
+                    }
+                }
+
+                if (binariesAreNewer) {
+                    JavaIndex.LOG.log(Level.FINE, "Using binaries for {0}", FileUtil.getFileDisplayName(root)); // NOI18N
+                    File copyTo = JavaIndex.getClassFolder(context);
+                    at = 0;
+                    for (URL singleBinaryRoot : binaryRoots) {
+                        FileObject singleBinaryRootFo = binaryRootsFo[at++];
+                        JavaBinaryIndexer.doIndex(context, singleBinaryRoot);
+                        JavaIndex.LOG.log(Level.FINE, "  copying from {0} to {1}", new Object[] { FileUtil.getFileDisplayName(singleBinaryRootFo), copyTo }); // NOI18N
+                        Enumeration<? extends FileObject> en = singleBinaryRootFo.getChildren(true);
+                        while (en.hasMoreElements()) {
+                            FileObject ch = en.nextElement();
+                            if (!ch.isData()) {
+                                continue;
+                            }
+                            String path = FileUtil.getRelativePath(singleBinaryRootFo, ch.getParent());
+                            if (path == null) {
+                                continue;
+                            }
+                            String name = ch.getNameExt();
+                            if (ch.hasExt("class")) {
+                                name = ch.getName() + ".sig";
+                            }
+                            File toDir = new File(copyTo, path.replace('/', File.separatorChar));
+                            toDir.mkdirs();
+                            File to = new File(toDir, name);
+                            try (OutputStream os = new FileOutputStream(to); InputStream is = ch.getInputStream()) {
+                                FileUtil.copy(is, os);
+                            }
+                        }
+                    }
```

Comment:
I think that `APTUtils.sourceRootRegistered(context.getRoot(), context.getRootURI());` should be called here.

Reply:
Is c1dc05a8138b correct?